qiskit/circuit/instructionset.py

```diff
@@ -13,28 +13,132 @@
+            if classical not in clbit_set:
+                raise CircuitError(
+                    f"Condition bit {classical} is not in the registers known here: {creg_set}"
+                )
+            return classical
+        if isinstance(classical, ClassicalRegister):
+            if classical not in creg_set:
+                raise CircuitError(
+                    f"Condition register {classical} is not one of the registers known here:"
+                    f" {creg_set}"
+                )
+            return classical
+        if isinstance(classical, int):
+            try:
+                return clbit_flat[classical]
+            except IndexError:
+                raise CircuitError(f"Bit index {classical} is out-of-range.") from None
+        raise CircuitError(
+            "Invalid classical condition. Must be an int, Clbit or ClassicalRegister, but received"
+            f" '{classical}'."
+        )
+
+    return resolver
 
 
 class InstructionSet:
     """Instruction collection, and their contexts."""
 
-    def __init__(self, circuit_cregs=None):
+    __slots__ = ("instructions", "qargs", "cargs", "_resolver")
+
+    def __init__(self, circuit_cregs=None, *, resource_resolver: Optional[Callable] = None):
         """New collection of instructions.
 
-        The context (qargs and cargs that each instruction is attached to)
-        is also stored separately for each instruction.
+        The context (qargs and cargs that each instruction is attached to) is also stored separately
+        for each instruction.
 
         Args:
             circuit_cregs (list[ClassicalRegister]): Optional. List of cregs of the
                 circuit to which the instruction is added. Default: `None`.
+
+                .. deprecated:: qiskit-terra 0.19
+                    The classical registers are insufficient to access all classical resources in a
+                    circuit, as there may be loose classical bits as well.  It can also cause
+                    integer indices to be resolved incorrectly if any registers overlap.  Instead,
+                    pass a complete resolver to the ``resource_resolver`` argument.
+
+            resource_resolver: A callable that takes in the classical resource used in the
+                condition, verifies that it is present in the attached circuit, resolves any indices
+                into concrete :obj:`.Clbit` instances, and returns the concrete resource.  If this
+                is not given, specifying a condition with an index is forbidden, and all concrete
+                :obj:`.Clbit` and :obj:`.ClassicalRegister` resources will be assumed to be valid.
+
+                .. note::
+
+                    The callback ``resource_resolver`` is called once for each call to
+                    :meth:`.c_if`, and is allowed to assume that a call implies that the resource
+                    will now be used.  It may throw an error if the resource is not valid for usage.
```

Comment:
Why would the `resource_resolver` need (or want) to assume a call implies the resource will be used?

Reply:
It's 100% for the control-flow builder interface.  The call to `InstructionSet.c_if` occurs after the call to `QuantumCircuit.append`, which is where the scoping constructs hook into, so normally they wouldn't be able to track any additional classical resources that are inserted by a `c_if` call.  With this set-up, once you enter a scope, the resolver for the `InstructionSet` now comes from the scope instead, which lets it track injected resources.

Perhaps the name of the function is wrong here - it's not _just_ a resolver, it's actually resolves, verifies and tracks resources that are given to it.

Reply:
Got it, okay. Maybe in the context of the builder PR, a better name might come to mind, but I think for now this is good as is.

Reply:
Perhaps `request_resource` instead of `resolve_resource`?  That has the right sort of implications in the name, I think.

Reply:
Changed to `resource_requester` in f8ce816dd.